### PR TITLE
MAINT: TST: Skip tests with multiprocessing that use "spawn" start method

### LINKED ISSUE #11033: deadlock on osx for python 3.8
We are still seeing something weird with a python 3.8 build on osx over on conda-forge

```
==== testing package: scipy-1.3.1-py38h82752d6_2 =====
running run_test.py
============================= test session starts ==============================
platform darwin -- Python 3.8.0, pytest-5.2.2, py-1.8.0, pluggy-0.12.0 -- $PREFIX/bin/python
cachedir: .pytest_cache
rootdir: $SRC_DIR
collecting ... collected 17186 items / 1225 deselected / 15961 selected

_lib/tests/test__gcutils.py::test_set_gc_state PASSED                    [  0%]
_lib/tests/test__gcutils.py::test_gc_state PASSED                        [  0%]
_lib/tests/test__gcutils.py::test_assert_deallocated PASSED              [  0%]
_lib/tests/test__gcutils.py::test_assert_deallocated_nodel PASSED        [  0%]
_lib/tests/test__gcutils.py::test_assert_deallocated_circular PASSED     [  0%]
_lib/tests/test__gcutils.py::test_assert_deallocated_circular2 PASSED    [  0%]
_lib/tests/test__testutils.py::test__parse_size PASSED                   [  0%]
_lib/tests/test__testutils.py::test__mem_available PASSED                [  0%]
_lib/tests/test__threadsafety.py::test_parallel_threads PASSED           [  0%]
_lib/tests/test__threadsafety.py::test_reentering PASSED                 [  0%]
_lib/tests/test__util.py::test__aligned_zeros PASSED                     [  0%]
_lib/tests/test__util.py::test_check_random_state PASSED                 [  0%]
_lib/tests/test__util.py::TestMapWrapper::test_serial PASSED             [  0%]
```

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=91407&view=logs

This is with a patch from the 1.3.x branch here 

https://github.com/conda-forge/scipy-feedstock/blob/02371e5393763b42fef7f50552829e57f2eb79f3/recipe/scipy1.3.2.patch

@rgommers  for viz

relates to https://github.com/conda-forge/scipy-feedstock/pull/114

### PR DESCRIPTION
**Pull request by ilayn**
#### Reference issue
Closes #11033 

#### What does this implement/fix?

Skips the multiprocessing tests unless the fork method is `'fork'`. Windows (all Python versions) and macOS (Python 3.8) use `'spawn'`, which hits an infinite recursion issue due to `pytest` + `setuptools` + `multiprocessing`.

I am not sure why this wasn't hit before at all.

### CODE REVIEW

File: scipy/_lib/tests/test__util.py [hunk 1]
```diff
@@ -85,37 +85,39 @@ def test_serial(self):
...
-            assert_(q._own_pool is False)
-            q.close()
-
-            # closing the PoolWrapper shouldn't close the internal pool
-            # because it didn't create it
-            out = p.map(np.sin, self.input)
-            assert_equal(list(out), self.output)
-        finally:
-            p.close()
+        # This is needed for Windows to get out of infinite recursion
```
Comment (WarrenWeckesser) [inline scipy/_lib/tests/test__util.py]: Not just Windows.  The problem also occurs on Mac OSX with Python 3.8.

Comment (ilayn) [reply scipy/_lib/tests/test__util.py]: Indeed I'll fix it once I see the green tick and merge right away.

**Comment by tylerjereddy** [comment]
If it needs to be in `1.3.3` can you add a backport label? My impression from the discussion in the issue is that this is an adjustment to improve test suite behavior, but not a serious bug in actual library code, right?

**Comment by ilayn** [comment]
Yes this affects only the test behavior and code is fine. While this is good news for us, unfortunately because it chokes many builds downstream, we might need to backport it. 

Sorry for not working on it a bit earlier. Took me a while to figure it out.

**Comment by ilayn** [comment]
The failing test `test_mstats_basic.py::TestCompareWithStats::test_kendalltau` doesn't look like related. So I'll ignore it. But AppVeyor queue is apparently clogged and queue is at least 4 hours.

**Comment by WarrenWeckesser** [comment]
Update to my [comment in the issue](https://github.com/scipy/scipy/issues/11033#issuecomment-552116498): with this PR, `../../tests/test__util.py::TestMapWrapper::test_parallel` now runs, but the test suite now appears to hang at

    ../../fft/tests/test_multithreading.py::test_mixed_threads_processes

**Comment by tylerjereddy** [comment]
Ok, sounds good--once this stuff is fully resolved we can put out another 1.3.x release. I think Ralf was "ok" with getting rid of Appveyor in our CI at this point to be honest.

File: scipy/_lib/tests/test__util.py [hunk 2]
```diff
@@ -85,37 +85,41 @@ def test_serial(self):
...
-            # closing the PoolWrapper shouldn't close the internal pool
-            # because it didn't create it
-            out = p.map(np.sin, self.input)
-            assert_equal(list(out), self.output)
-        finally:
-            p.close()
+        # The line __name__ == '__main__' is needed for platforms that
+        # uses spawn() instead of fork() to start new processes, to get
+        # out of infinite recursion.
+        if __name__ == '__main__':
```
Comment (WarrenWeckesser) [inline scipy/_lib/tests/test__util.py]: Under what conditions is this condition true?  Have you, in effect, simply commented-out the test, for all platforms?

**Comment by mdhaber** [comment]
Same for me as @WarrenWeckesser; now `scipy\fft\tests\test_multithreading.py::test_mixed_threads_processes` hangs.
(Testing on Windows w/ conda Python 2.7 and openblas built from source)

File: scipy/_lib/tests/test__util.py (hunk 2 above)
Comment (ilayn) [reply scipy/_lib/tests/test__util.py]: This condition is true only for the first time this instance is created, then for other processes the name is not main anymore. Hence it blocks the recursion.

Comment (WarrenWeckesser) [reply scipy/_lib/tests/test__util.py]: My understanding is that `__name__` is `__main__` when the file is run as a script.  Does `pytest` run the file as a script?  I just added a couple lines with `assert_(False)` to the test, and it still passed.  I.e. the code in the `with` context never ran.

**Comment by ilayn** [comment]
Ah missed that one. Let me do a local full run and add other potential misses.

**Comment by WarrenWeckesser** [comment]
As far as I can tell, this change is not correct.  I don't think `pytest` will ever run a file such that `__name__` has the value `'__main__'`, so all this change does is comment-out the test on all platforms.  You can test this by adding, say, `assert_(False, "always fail!")` somewhere in the code that should be run when `__name__ == '__main__'`.  In my tests with Python 3.7 and 3.8 on Mac OS X 10.12.6, that code is never run.

**Comment by ilayn** [comment]
You might be right. I did the same thing but somehow it might have to do with the fact that when I isolated them in a separate script .py file which probably made it `__main__` internally.

**Comment by andyfaff** [comment]
Something smells here. I could vaguely see why incrementing major Python would start causing issues. However, I don't understand why the test suite has now started failing on Python < 3.8. e.g:

> scipy\fft\tests\test_multithreading.py::test_mixed_threads_processes hangs. (Testing on Windows w/ conda Python 2.7 and openblas built from source)

The test suite has been running fine up until now.

**Comment by ilayn** [comment]
@WarrenWeckesser That's a good catch indeed. You are definitely right. We have to switch to what @rgommers suggested then ☹️

**Comment by ilayn** [comment]
And also I'm out of ideas how to modify the test script such that it executes a `python -m pytest` instead of the current `pytest.main()`

**Comment by rgommers** [comment]
> And also I'm out of ideas how to modify the test script such that it executes a `python -m pytest` instead of the current `pytest.main()`

I doubt this is going to be robust anyway, there's too many ways of running the tests. Quick fix is to skip the tests on Windows and macOS.

**Comment by larsoner** [comment]
Could we skip based on the forking method?

FWIW we might also be able in our `runtests.py` set the forking method, if that's the problem.

**Comment by andyfaff** [comment]
I use almost exactly the same MapWrapper code in my own project, with similar tests. I use `python setup.py test` and `pytest` to start the test runs. I don't see any issues with py38 on macOS.

One small difference is that `get_context` is used to change between fork/spawn.

**Comment by ilayn** [comment]
@andyfaff From what I have read, that small difference is the root cause on windows machines. As I wrote in the issue, I don't know why it started to choke suddenly.

**Comment by ilayn** [comment]
I've spared a few more hours to this. As far as I can read, almost unanimously accepted that, there is no sensible entry point for pytest injecting statements for `spawn` method. And it goes all the way down the rabbit hole to setuptools' entry points.

The other alternative is changing the whole behavior of `runtests.py` such that it is fired from another subprocess which doesn't seem sensible to me at this stage.

If there is noone who understands about this, I propose to skip these tests with links to this discussion.

**Comment by rgommers** [comment]
Thanks @ilayn, I agree with skipping the tests. Skiping on
```
multiprocessing.get_start_method() is not 'fork'
```
seems reasonable.

**Comment by ilayn** [comment]
Tested and skipped tests locally, then rebased. Also separated the tests for later convenience since there were only two lines in common.

**Comment by rgommers** [comment]
Looks like the `optimize.brute` parallel test needs to be skipped in the same manner, see https://github.com/conda-forge/scipy-feedstock/pull/113#issuecomment-552204981. See `scipy-1.3.2-bad-tests.patch` in that PR to compare if we've got them all here.

**Comment by ilayn** [comment]
Ah OK. let me add that. But I still wonder why only one of the runs in Azure shows the actual skip but somehow all pass. Lately I have the impression (more like a hunch) that not all tests are run.

**Comment by rgommers** [comment]
> Lately I have the impression (more like a hunch) that not all tests are run.

That wouldn't be the first time, it's really hard to figure out though. With `nose` this also happened semi-regularly, due to file naming, permissions on files, or some random voodoo.

**Comment by rgommers** [comment]
I had the same in another PR just now - it added f-strings that were covered by tests, but somehow the py35 build matrix entry didn't complain about syntax errors.

**Comment by tylerjereddy** [comment]
Hmm, coverage reporting is useful for probing that--tests should usually be fully executed except for skips. Unfortunately coverage reporting is an ecosystem weakness as well.

**Comment by ilayn** [comment]
I'm on the phone so it is a bit difficult for me so can someone restart the final test on Travis please?

**Comment by rgommers** [comment]
All green and LGTM, merging. Thanks @ilayn

**Comment by rgommers** [comment]
for next time, better not mix in lots of style cleanups - that makes it much harder to review

**Comment by tylerjereddy** [comment]
`1.3.3` milestone I guess so that 3.8 tests behave well before December release of 1.4.0.

**Comment by rgommers** [comment]
The workflow we're using (I think?) is: when merged, use the milestone it's in. Label with `backport-candidate` if appropriate. If backported, then change the milestone and remove the label.

**Comment by tylerjereddy** [comment]
alright

**Pull request merged**